connectors/routing/routing_config.go

```diff
@@ -21,57 +21,165 @@
 package routing
 
 import (
+	"fmt"
+	"regexp"
 	"strings"
 
-	"github.com/gobwas/glob"
 	"github.com/pkg/errors"
+	"github.com/uber-go/dosa"
 )
 
-// Rule stores routing rule
-// to decide which connector the Connector talks to
+const lastASCIIChar = "~"
+
+// Rule is an assignment from scope.prefixPattern to a connector name.
 type Rule struct {
-	Scope      string
-	NamePrefix string
-	Connector  string
-	GlobMatch  glob.Glob
+	scope           string
+	namePrefix      string
+	connector       string
+	prefixIsPattern bool
+	prefixPattern   *regexp.Regexp
+	scopeIsPattern  bool
+	scopePattern    *regexp.Regexp
+	canonScope      string
+	canonPfx        string
 }
 
 // NewRule initializes Rule
 func NewRule(scope, namePrefix, connector string) (*Rule, error) {
+	// scope must be a valid name, optionally with a suffix *.
+	if scope == "" {
+		return nil, errors.New("scope cannot be empty")
+	}
+	var scopeIsPattern bool
+	if strings.HasSuffix(scope, "*") {
+		scopeIsPattern = true
+		scope = strings.TrimSuffix(scope, "*")
+	}
+	nScope, err := canonicalize(scope, scopeIsPattern, true) // isScope = true
+	if err != nil {
+		return nil, errors.Wrap(err, fmt.Sprintf("invalid scope name %s", scope))
+	}
+
+	// namePrefix must be a valid prefix name, optionally with a suffix *
 	if namePrefix == "" {
-		return nil, errors.New("namePrefix could not be empty, should be defined in yaml file")
+		return nil, errors.New("namePrefix cannot be empty")
+	}
+	var prefixIsPattern bool
+	if strings.HasSuffix(namePrefix, "*") {
+		prefixIsPattern = true
+		namePrefix = strings.TrimSuffix(namePrefix, "*")
+	}
+	if namePrefix != "" {
+		if err := dosa.IsValidNamePrefix(namePrefix); err != nil {
+			return nil, errors.Wrap(err, fmt.Sprintf("could not parse routing rule: invalid namePrefix %s", namePrefix))
+		}
 	}
+	nPrefix, _ := canonicalize(namePrefix, prefixIsPattern, false) // isScope = false
 
-	if scope == "" {
-		return nil, errors.New("scope could not be empty, should be defined in yaml file")
+	// Make a regular expressions and matchers
+	prefixPat := makeRegexp(namePrefix, prefixIsPattern)
+	scopePat := makeRegexp(scope, scopeIsPattern)
+
+	return &Rule{
+		namePrefix:      namePrefix,
+		scope:           scope,
+		connector:       connector,
+		scopeIsPattern:  scopeIsPattern,
+		scopePattern:    scopePat,
+		prefixIsPattern: prefixIsPattern,
+		prefixPattern:   prefixPat,
+		canonScope:      nScope,
+		canonPfx:        nPrefix,
+	}, nil
+}
+
+// RouteTo (bad name!) says whether or not this Rule can handle the given scope:prefix.
```

Comment:
why don't we change the name if it's a bad name

Reply:
A diff is not ready for reviewing until I add reviewers.....